### DISCUSSION TITLE
Using crane with wasm-pack

### DISCUSSION THREAD

**Discussion by yannham** [ipetkov/crane#193] [Q&A] [answered]
Hello,

We've moved away from `import-cargo` to `crane` for [Nickel](https://github.com/tweag/nickel) with success (in particular, the documentation of crane has been very appreciated, thanks!).

In a second time, we would like to use crane to build the WASM REPL as well. Nothing fancy, it's a wrapper crate around the main one, and we must produce a package accessible from JavaScript. However the current build relies on the `wasm-pack` tool, and the naive change (substitute `import-cargo`'s `cargoHome` for to the one generated by crane in the build dependencies) doesn't work. The build tries to access the network. See the tentative change here: https://github.com/tweag/nickel/pull/1014.

We're going to investigate more, but in the meantime, is there any obvious solution, or something that is done obviously wrong?

Thanks!

**Comment by ipetkov** [comment, marked as answer]
Hi @yannham glad you are appreciating the documentation!

It's been a while since I've last used `wasm-pack` (especially since the last time I tried using it with Nix too) and I do remember hitting some headaches from it trying to access the network from inside the sandbox. Although it doesn't seem to support an offline mode, I believe it will skip trying to install various tools if they're already on the path (IIRC it needs `pkgs.binaryen` and `pkgs.wasm-bindgen-cli` on the `PATH`)

Assuming that fixes any sandbox related issues, I think it should be possible to make a build work. You might have to drop down to using `mkCargoDerivation`, here's a quick and dirty example:

```nix
{
  inputs = {
    nixpkgs.url = "nixpkgs/nixpkgs-unstable";
    crane = {
      url = "github:ipetkov/crane";
      inputs.nixpkgs.follows = "nixpkgs";
    };
    rust-overlay = {
      url = "github:oxalica/rust-overlay";
      inputs.nixpkgs.follows = "nixpkgs";
    };
    utils = {
      url = "github:numtide/flake-utils";
      inputs.nixpkgs.follows = "nixpkgs";
    };
  };

  outputs = inputs@{ self, nixpkgs, utils, ... }:
    utils.lib.eachDefaultSystem (system:
      let
        pkgs = import nixpkgs {
          inherit system;
          overlays = [
            (import inputs.rust-overlay)
          ];
        };

        rust-toolchain = pkgs.rust-bin.stable.latest.default.override {
          targets = [
            "wasm32-unknown-unknown"
          ];
        };

        craneLib = (inputs.crane.mkLib pkgs).overrideToolchain rust-toolchain;

        src = craneLib.cleanCargoSources ./.;
        cargoArtifacts = craneLib.buildDepsOnly {
          inherit src;
          cargoExtraArgs = "--target wasm32-unknown-unknown";
          doCheck = false;
        };
      in
      rec {
        packages.default = craneLib.mkCargoDerivation {
          inherit cargoArtifacts src;
          buildPhaseCargoCommand = ''
            wasm-pack build
          '';
          installPhaseCommand = ''
            # Install whatever is needed
            mv pkg $out
          '';
          nativeBuildInputs = with pkgs; [
            binaryen
            wasm-bindgen-cli
            wasm-pack
          ];
        };
      }
    );
}
```

**Reply by yannham** [reply]
Thank you for the quick reply.

I'm making progress! `wasm-pack` now starts to run and to compile the dependencies, but for some reason, it fails with "Permission denied (os error 13)` on custom build scripts of said dependencies (serde, typenum, ... there doesn't seem to be a pattern here, any custom build script is failing). Any clue on what could be the issue?

```
nickel-lang> [Nix build script] Running wasm-pack
nickel-lang> [INFO]: Checking for the Wasm target...
nickel-lang> [INFO]: Compiling to Wasm...
nickel-lang>    Compiling syn v1.0.95
nickel-lang>    Compiling cfg-if v1.0.0
nickel-lang>    Compiling typenum v1.15.0
nickel-lang>    Compiling serde v1.0.137
nickel-lang> error: failed to run custom build command for `typenum v1.15.0`
nickel-lang> Caused by:
nickel-lang>   could not execute process `/build/source/target/release/build/typenum-6ba44e225a51239a/build-script-main` (never executed)
nickel-lang> Caused by:
nickel-lang>   Permission denied (os error 13)
nickel-lang> warning: build failed, waiting for other jobs to finish...
nickel-lang> error: failed to run custom build command for `serde v1.0.137`
nickel-lang> Caused by:
nickel-lang>   could not execute process `/build/source/target/release/build/serde-01a59449dfd038a1/build-script-build` (never executed)
nickel-lang> Caused by:
nickel-lang>   Permission denied (os error 13)
nickel-lang> Error: Compiling your crate to WebAssembly failed
nickel-lang> Caused by: failed to execute `cargo build`: exited with exit status: 101
nickel-lang>   full command: "cargo" "build" "--lib" "--release" "--target" "wasm32-unknown-unknown" "--no-default-features" "--frozen" "--offline"
error: builder for '/nix/store/skjkb88r4swpc70l21g74xkv5brmi6vb-nickel-lang-0.3.1.drv' failed with exit code 1;
       last 10 log lines:
       > error: failed to run custom build command for `serde v1.0.137`
       >
       > Caused by:
       >   could not execute process `/build/source/target/release/build/serde-01a59449dfd038a1/build-script-build` (never executed)
       >
       > Caused by:
       >   Permission denied (os error 13)
       > Error: Compiling your crate to WebAssembly failed
       > Caused by: failed to execute `cargo build`: exited with exit status: 101
       >   full command: "cargo" "build" "--lib" "--release" "--target" "wasm32-unknown-unknown" "--no-default-features" "--frozen" "--offline"
       For full logs, run 'nix log /nix/store/skjkb88r4swpc70l21g74xkv5brmi6vb-nickel-lang-0.3.1.drv'.
```

**Reply by ipetkov** [reply]
@yannham I took a quick peek at your WIP branch and I noticed a few things which might help:

1. the `buildDepsOnly` derivation isn't actually building for `wasm32-unknown-unknown` but defaulting to whatever the host platform is. You'll likely want to pass `--target wasm32-unknown-unknown` to cargo, otherwise the actual wasm-pack build won't reuse any of those artifacts (and end up compiling them all again)
2. Seems `--workspace` is being set in your configuration somehow. It seems like there are some dependencies or crates in the overall workspace which can't be built for wasm. The `buildDepsOnly` derivation will probably want to pass `-p nickel-repl` to cargo to only build the crate you care about
3. Might also want to set `doCheck = false;` in the `buildDepsOnly` derivation since cargo can't run any tests on its own (unless you override the default checks with your own test harness, but I'll leave that up to you)

Hope this helps!

**Reply by yannham** [reply]
Hooray, this did the trick :tada: Thanks a lot. I initially removed the `--target wasm32-unknown-unknwon` because I'm supposed to provide a rust toolchain with target set to "wasm32-unknown-unknown" already, I thought it would be sufficient. But indeed if I remove it it fails again. This required a bit of specific knowledge, but the resulting flake looks rather clean and natural!

**Reply by ipetkov** [reply]
> because I'm supposed to provide a rust toolchain with target set to "wasm32-unknown-unknown" already, I thought it would be sufficient

Ah I see, so just to clarify: adding a particular target to the rust toolchain means that that target _becomes available for use_ but doesn't necessarily replace the platform specific target (I mean, technically I think you _could_ do that using rustup, but I think `rust-overlay`'s behavior is additive-only)

**Comment by hsjobeki** [comment]
@ipetkov have been stuck on this.
> error: builder for '/nix/store/skjkb88r4swpc70l21g74xkv5brmi6vb-nickel-lang-0.3.1.drv' failed with exit code 1;
       last 10 log lines:
       > error: failed to run custom build command for `serde v1.0.137`
       >
       > Caused by:
       >   could not execute process `/build/source/target/release/build/serde-01a59449dfd038a1/build-script-build` (never executed)
       >
       > Caused by:
       >   Permission denied (os error 13)
       
It seems `wasm-pack` tries to read the `HOME` directory, which fails inside the nix-sandbox.

> [INFO]: :-) Your wasm pkg is ready to publish at /home/johannes/git/tmp/wasm-bindgen-example/pkg.

I fixed this by setting $HOME='.' before the build

```
buildPhaseCargoCommand = ''
  export HOME=.
  wasm-pack build
'';
```

When leaving $HOME as default or setting it to /homeless-shelter wasm build fails.

We could add that as an example as well. Because not everyone builds web apps with `trunk`.

**Reply by ipetkov** [reply]
I know wasm-pack tries to download stuff sometimes, but it might be trying to check `${HOME}/.cache`. Maybe `XDG_CACHE_HOME` needs to be set as an alternative...

Happy to review a documentation PR to improve the examples though!
